design/tla/Task.tla

```diff
@@ -0,0 +1,760 @@
+                 ] @@ tasks
+
+(* Actions we require to happen eventually when possible. *)
+AgentProgress ==
+  \/ ProgressTask
+  \/ ShutdownComplete
+
+(* All actions of the agent/worker. *)
+Agent ==
+  \/ AgentProgress
+  \/ RejectTask
+  \/ ContainerExit
+  \/ WorkerDown
+
+-------------------------------------------------------------------------------
+\*  Actions performed by the reaper
+
+(* Forget about tasks in remove or orphan states.
+
+   XXX: really forget about orphans? *)
+Reaper ==
+  /\ UNCHANGED << services, nEvents >>
+  /\ \E i \in DOMAIN tasks :
+      /\ State(i) \in {remove, orphaned}
+      /\ tasks' = [t \in DOMAIN tasks \ {i} |-> tasks[t]]
+
+-------------------------------------------------------------------------------
+\*  The complete system
+
+\* All the variables
+vars == << tasks, services, nEvents >>
+
+\* Initially there are no tasks and no services
+Init ==
+  /\ tasks = << >>
+  /\ services = << >>
+  /\ InitEvents
+
+(* A next step is one in which any of these sub-components takes a step: *)
+Next ==
+  \/ User
+  \/ Orchestrator
+  \/ Allocator
+  \/ Scheduler
+  \/ Agent
+  \/ Reaper
+  \* For model checking: don't report deadlock if we're limiting events
+  \/ (nEvents = maxEvents /\ UNCHANGED vars)
+
+(* This is a ``temporal formula''. It takes a sequence of states representing the
+   changing state of the world and evaluates to TRUE if that sequences of states is
+   a possible behaviour of SwarmKit. *)
+Spec ==
+  \* The first state in the behaviour must satisfy Init:
+  /\ Init
+  \* All consecutive pairs of states must satisfy Next or leave `vars' unchanged:
+  /\ [][Next]_vars
+  (* Some actions are required to happen eventually. For example, a behaviour in
+     which SwarmKit stops doing anything forever, even though it could advance some task
+     from the `new' state, isn't a valid behaviour of the system.
+     This property is called ``weak fairness''. *)
+  /\ WF_vars(OrchestratorProgress \/ Allocator \/ Scheduler \/ AgentProgress \/ Reaper)
+     (* We don't require fairness of:
+        - User (we don't control them),
+        - RestartTask (services aren't required to be updated),
+        - RejectTask (tasks aren't required to be rejected),
+        - ContainerExit (we don't specify image behaviour) or
+        - WorkerDown (workers aren't required to fail). *)
+
+-------------------------------------------------------------------------------
+\* Properties to verify
+
+(* These are properties that should follow automatically if the system behaves as
+   described by `Spec' in the previous section. *)
+
+\* A state invariant (things that should be true in every state).
+Inv ==
+  \A id \in DOMAIN tasks :
+    (* Every task has a service:
+
+       XXX: The spec says: ``In some cases, there are tasks that exist independent of any service.
+            These do not have a value set in service_id.''. Need an example of one. *)
```

Comment:
Network attachment tasks. They're an abomination, an absolutely disgusting hack that should never have been shipped. They are tasks which use the `NetworkAttachmentSpec` runtime spec, a spec which contains a container ID. These tasks are created through the grpc server implemented in `resourceapi`, and they are created _by the agent_. The purpose of Network Attachment tasks is to attach regular `docker run` containers to a swarmkit overlay network that has been created with the `--attachable` option.

There have been bugs in the past with attachment tasks because we have no visibility into their lifecycle and their creation and destruction is entirely at the whim of the agent. The end result is that network attachment tasks would remain in the object store after they had been terminated, leaking resources and causing the associated attached networks to be unremovable.

I added a workaround to remove network attachment tasks when a node is removed in #2409, but the underlying issue IIRC remains.

I'm sorry.

Reply:
Also, @jlhawn is working on a hack week project to run one-off tasks, like `docker run`-esque tasks, that would exist independently of services.